src/extension/authkey/src/main/java/org/geoserver/security/GeoServerAuthenticationKeyFilter.java

```diff
@@ -68,26 +71,27 @@ public void initializeFromConfig(SecurityNamedServiceConfig config) throws IOExc
         mapper.setUserGroupServiceName(userGroupServiceName);
         mapper.setSecurityManager(getSecurityManager());
         mapper.configureMapper(authConfig.getMapperParameters());
+        allowChallengeAnonymousSessions = authConfig.isAllowChallengeAnonymousSessions();
     }
 
     @Override
     public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
             throws IOException, ServletException {
 
-        // String authKey = getAuthKey((HttpServletRequest) request);
-        // if (authKey==null) { // nothing to do
-        // chain.doFilter(request, response);
-        // return;
-        // }
-
-        String cacheKey = authenticateFromCache(this, (HttpServletRequest) request);
+        String cacheKey =
+                authenticateFromCache(
+                        this, (HttpServletRequest) request, allowChallengeAnonymousSessions);
 
-        if (SecurityContextHolder.getContext().getAuthentication() == null) {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        if (allowChallengeAnonymousSessions
+                || authentication == null
+                || authentication instanceof AnonymousAuthenticationToken) {
```

Comment:
It seems we get into this branch even if it's just an anonymous auth, even whenallowChallengeAnonymousSessions is false... was that intended?

Reply:
Yes, it was intended. If the filter chain hits the default one, the latter injects an `AnonymousAuthenticationToken` into the `SecurityContext` and we must threat this one the same as it was not authenticated (`authentication == null`).